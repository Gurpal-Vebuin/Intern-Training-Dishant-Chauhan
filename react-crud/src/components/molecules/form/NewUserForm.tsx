import React from "react";
import { InputValue } from "../../atoms/Input/InputValue";
import useLanguage from "../../../hooks/useLanguage";
import { FormWrapper } from "../../styles/form/FormStyles";
import { useForm } from "react-hook-form";
import { joiResolver } from "@hookform/resolvers/joi";
import { registerSchema } from "../../../data/models/registerSchema";
import { ErrorMessage, LabelWrapper } from "../../styles/form/InputStyles";
import { StyledOption, StyledSelect } from "../../styles/select/SelectStyles";
import { NewFormProps, FormData } from "../../../types/interfaces/interface";
import { useNavigate } from "react-router-dom";
import { Links } from "../../../types/constants/Links/Links";

const NewUserForm: React.FC<NewFormProps> = ({ onSubmit }) => {
  const { t } = useLanguage();
  const navigate = useNavigate();

  const {
    register,
    handleSubmit,
    formState: { errors },
    setValue,
    trigger,
    watch,
  } = useForm<FormData>({
    resolver: joiResolver(registerSchema),
    defaultValues: {
      name: "",
      email: "",
      phone: "",
      password: "",
      roles: "user",
    },
  });

  const selectedRole = watch("roles");

  return (

      <FormWrapper onSubmit={handleSubmit(onSubmit)}>
        <h2>{t("Register")}</h2>

        <InputValue
          htmlFor="name"
          type="text"
          label={t("Form.Name")}
          id="name"
          placeholder={t("NamePlaceholder")}
          registerProps={register("name")}
          />

        {errors.name && <ErrorMessage>{errors.name.message}</ErrorMessage>}

        <InputValue
          htmlFor="email"
          type="email"
          label={t("Form.Email")}
          id="email"
          placeholder={t("EmailPlaceholder")}
          registerProps={register("email")}
          />
        {errors.email && <ErrorMessage>{errors.email.message}</ErrorMessage>}

        <InputValue
          htmlFor="phone"
          type="text"
          label={t("Form.Phone")}
          id="phone"
          placeholder={t("PhonePlaceholder")}
          registerProps={register("phone")}
          />
        {errors.phone && <ErrorMessage>{errors.phone.message}</ErrorMessage>}

        <InputValue
          htmlFor="password"
          type="password"
          label={t("Form.Password")}
          id="password"
          placeholder={t("PasswordPlaceholder")}
          registerProps={register("password")}
          />
        {errors.password && (
          <ErrorMessage>{errors.password.message}</ErrorMessage>
        )}

        <LabelWrapper htmlFor="roles">{t("Form.Roles")}</LabelWrapper>
        <StyledSelect
          id="roles"
          {...register("roles")}
          value={selectedRole}
          onChange={(e) => {
            setValue("roles", e.target.value);
            trigger("roles");
          }}
          >
          <StyledOption value="user">{t("Form.User")}</StyledOption>
          <StyledOption value="admin">{t("Form.Admin")}</StyledOption>
        </StyledSelect>
        {errors.roles && <ErrorMessage>{errors.roles.message}</ErrorMessage>}

        <div>
          <button type="submit">{t("Register")}</button>
          <p>
            {t("Registered")}{" "}
            <span onClick={() => navigate(`${Links.HOME}`)}>{t("Login")}</span>
          </p>
        </div>
      </FormWrapper>
    
  );
};

export default NewUserForm;
