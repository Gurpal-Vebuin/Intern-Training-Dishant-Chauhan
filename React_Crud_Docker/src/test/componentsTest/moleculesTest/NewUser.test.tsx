import "@testing-library/jest-dom";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { useNavigate } from "react-router-dom";
import { useForm } from "react-hook-form";
import { Links } from "../../../types/constants/Links/Links";
import NewUserForm from "../../../components/molecules/form/NewUserForm";

jest.mock("react-router-dom", () => ({
  useNavigate: jest.fn(),
}));

jest.mock("../../../hooks/useLanguage", () => ({
  __esModule: true,
  default: jest.fn(() => ({ t: (key: string) => key })),
}));

jest.mock("react-hook-form", () => ({
  __esModule: true,
  useForm: jest.fn(),
}));

describe("NewUserForm Component", () => {
  const mockNavigate = jest.fn();
  const mockOnSubmit = jest.fn();

  beforeEach(() => {
    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);

    (useForm as jest.Mock).mockReturnValue({
      register: jest.fn(),
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      handleSubmit: (fn: any) => fn,
      formState: { errors: {} },
      setValue: jest.fn(),
      trigger: jest.fn(),
      watch: jest.fn(() => "user"),
    });

    jest.clearAllMocks();
  });

  test("renders the form correctly", () => {
    render(<NewUserForm onSubmit={mockOnSubmit} />);

    expect(screen.getAllByText("Register").length).toBeGreaterThan(0);
    expect(screen.getByText("Form.Name")).toBeInTheDocument();
    expect(screen.getByText("Form.Email")).toBeInTheDocument();
    expect(screen.getByText("Form.Phone")).toBeInTheDocument();
    expect(screen.getByText("Form.Password")).toBeInTheDocument();
    expect(screen.getByText("Form.Roles")).toBeInTheDocument();

    expect(
      screen.getByRole("button", { name: /Register/i })
    ).toBeInTheDocument();
  });

  test("calls onSubmit function when form is valid", async () => {
    render(<NewUserForm onSubmit={mockOnSubmit} />);

    fireEvent.click(screen.getByRole("button", { name: /Register/i }));

    await waitFor(() => {
      expect(mockOnSubmit).toHaveBeenCalled();
    });
  });

  test("navigates to home page when clicking login link", () => {
    render(<NewUserForm onSubmit={mockOnSubmit} />);

    fireEvent.click(screen.getByText("Login"));

    expect(mockNavigate).toHaveBeenCalledWith(Links.HOME);
  });

  test("Creates a snapshot for Register Form", () => {
    const { asFragment } = render(
      <NewUserForm onSubmit={mockOnSubmit}></NewUserForm>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
