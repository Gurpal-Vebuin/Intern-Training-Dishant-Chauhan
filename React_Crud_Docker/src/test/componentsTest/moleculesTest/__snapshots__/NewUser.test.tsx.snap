// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewUserForm Component Creates a snapshot for Register Form 1`] = `
<DocumentFragment>
  <form
    class="sc-dycHyt ibuWyK"
  >
    <h2>
      Register
    </h2>
    <div
      class="sc-dVBluf gfXbFh"
    >
      <label
        class="sc-FEMpB dLTiLK"
        for="name"
      >
        Form.Name
      </label>
      <input
        aria-label="Form.Name"
        class="sc-gjcoXW swbNf"
        id="name"
        placeholder="NamePlaceholder"
        type="text"
      />
    </div>
    <div
      class="sc-dVBluf gfXbFh"
    >
      <label
        class="sc-FEMpB dLTiLK"
        for="email"
      >
        Form.Email
      </label>
      <input
        aria-label="Form.Email"
        class="sc-gjcoXW swbNf"
        id="email"
        placeholder="EmailPlaceholder"
        type="email"
      />
    </div>
    <div
      class="sc-dVBluf gfXbFh"
    >
      <label
        class="sc-FEMpB dLTiLK"
        for="phone"
      >
        Form.Phone
      </label>
      <input
        aria-label="Form.Phone"
        class="sc-gjcoXW swbNf"
        id="phone"
        placeholder="PhonePlaceholder"
        type="text"
      />
    </div>
    <div
      class="sc-dVBluf gfXbFh"
    >
      <label
        class="sc-FEMpB dLTiLK"
        for="password"
      >
        Form.Password
      </label>
      <input
        aria-label="Form.Password"
        class="sc-gjcoXW swbNf"
        id="password"
        placeholder="PasswordPlaceholder"
        type="password"
      />
    </div>
    <label
      class="sc-FEMpB dLTiLK"
      for="roles"
    >
      Form.Roles
    </label>
    <select
      aria-label="Role"
      class="sc-kfeOyU fjHYnM"
      id="roles"
    >
      <option
        class="sc-hRDKVd fidUmo"
        value="user"
      >
        Form.User
      </option>
      <option
        class="sc-hRDKVd fidUmo"
        value="admin"
      >
        Form.Admin
      </option>
    </select>
    <div>
      <button
        type="submit"
      >
        Register
      </button>
      <p>
        Registered 
        <span>
          Login
        </span>
      </p>
    </div>
  </form>
</DocumentFragment>
`;
